File: ports/zig/src/diff.zig
//! Turns two framebuffers into the smallest practical stream of escape
//! sequences. The encoder keeps a model of the terminal's current pen so no
//! redundant SGR is emitted.

const std = @import("std");

const ansi = @import("ansi.zig");
const buffer_mod = @import("buffer.zig");
const capabilities = @import("capabilities.zig");
const color_mod = @import("color.zig");
const unicode = @import("unicode.zig");

const Attrs = buffer_mod.Attrs;
const Color = color_mod.Color;
const ColorDepth = capabilities.ColorDepth;
const FrameBuffer = buffer_mod.FrameBuffer;

/// Rewriting up to this many unchanged cells is cheaper than the escape sequence
/// needed to jump over them, so neighbouring dirty spans get merged.
const gap_merge = 5;

pub const EncodeResult = struct {
    /// The bytes to hand to stdout. Owned by the encoder, valid until the next
    /// `encode`.
    output: []const u8,
    changed_cells: usize,
    /// Rows that contained at least one change.
    dirty_rows: usize,
};

pub const Options = struct {
    colors: ColorDepth = .truecolor,
    /// Drain all color, keeping attributes.
    monochrome: bool = false,
};

const State = struct {
    x: i64 = -1,
    y: i64 = -1,
    /// False after anything that makes the cursor position unknowable.
    known: bool = false,
    fg: Color = .default,
    bg: Color = .default,
    attrs: Attrs = .none,
};

pub const Encoder = struct {
    allocator: std.mem.Allocator,
    colors: ColorDepth,
    monochrome: bool,
    state: State = .{},
    parts: std.ArrayList(u8) = .empty,

    pub fn init(allocator: std.mem.Allocator, options: Options) Encoder {
        return .{
            .allocator = allocator,
            .colors = options.colors,
            .monochrome = options.monochrome,
        };
    }

    pub fn deinit(self: *Encoder) void {
        self.parts.deinit(self.allocator);
        self.* = undefined;
    }

    /// Forget what we believe about the terminal; the next write re-states
    /// everything.
    pub fn invalidateState(self: *Encoder) !void {
        self.state = .{};
        try self.parts.appendSlice(self.allocator, "\x1b[0m");
    }

    fn writeFg(self: *Encoder, c: Color) !void {
        if (self.colors == .none) return;
        if (c.isDefault()) {
            try self.parts.appendSlice(self.allocator, ansi.fg_default);
            return;
        }
        const col = if (self.monochrome) c.grayscale() else c;
        switch (self.colors) {
            .truecolor => try ansi.fgTrue(&self.parts, self.allocator, col.red(), col.green(), col.blue()),
            .ansi256 => try ansi.fg256(&self.parts, self.allocator, col.to256()),
            else => try ansi.fg16(&self.parts, self.allocator, col.to16()),
        }
    }

    fn writeBg(self: *Encoder, c: Color) !void {
        if (self.colors == .none) return;
        if (c.isDefault()) {
            try self.parts.appendSlice(self.allocator, ansi.bg_default);
            return;
        }
        const col = if (self.monochrome) c.grayscale() else c;
        switch (self.colors) {
            .truecolor => try ansi.bgTrue(&self.parts, self.allocator, col.red(), col.green(), col.blue()),
            .ansi256 => try ansi.bg256(&self.parts, self.allocator, col.to256()),
            else => try ansi.bg16(&self.parts, self.allocator, col.to16()),
        }
    }

    fn applyStyle(self: *Encoder, fg: Color, bg: Color, attrs: Attrs) !void {
        if (self.state.fg == fg and self.state.bg == bg and self.state.attrs.eql(attrs)) return;

        // Attributes can only be added cheaply; removing one means a full reset.
        if (!self.state.attrs.without(attrs).isEmpty()) {
            try self.parts.appendSlice(self.allocator, "\x1b[0m");
            self.state.attrs = .none;
            self.state.fg = .default;
            self.state.bg = .default;
        }

        const added = attrs.without(self.state.attrs);
        if (!added.isEmpty()) {
            var codes: [7]u8 = undefined;
            var n: usize = 0;
            inline for (.{
                .{ "bold", 1 },  .{ "dim", 2 },     .{ "italic", 3 },
                .{ "underline", 4 }, .{ "blink", 5 }, .{ "reverse", 7 },
                .{ "strike", 9 },
            }) |pair| {
                if (@field(added, pair[0])) {
                    codes[n] = pair[1];
                    n += 1;
                }
            }
            if (n > 0) {
                try self.parts.appendSlice(self.allocator, "\x1b[");
                for (codes[0..n], 0..) |code, i| {
                    if (i > 0) try self.parts.append(self.allocator, ';');
                    try self.parts.print(self.allocator, "{d}", .{code});
                }
                try self.parts.append(self.allocator, 'm');
            }
            self.state.attrs = attrs;
        }

        if (self.state.fg != fg) {
            try self.writeFg(fg);
            self.state.fg = fg;
        }
        if (self.state.bg != bg) {
            try self.writeBg(bg);
            self.state.bg = bg;
        }
    }

    fn moveCursor(self: *Encoder, x: usize, y: usize) !void {
        const xi: i64 = @intCast(x);
        const yi: i64 = @intCast(y);
        if (self.state.known and self.state.y == yi) {
            if (self.state.x == xi) return;
            if (xi > self.state.x and xi - self.state.x <= 3) {
                // Short hop: cheaper than a full CUP, and never repaints cells.
                try ansi.moveRight(&self.parts, self.allocator, @intCast(xi - self.state.x));
            } else if (x == 0) {
                try self.parts.append(self.allocator, '\r');
            } else {
                try ansi.moveToColumn(&self.parts, self.allocator, x);
            }
        } else {
            try ansi.moveTo(&self.parts, self.allocator, x, y);
        }
        self.state.x = xi;
        self.state.y = yi;
        self.state.known = true;
    }

    /// Encode the difference between `prev` and `next`. Pass `full` to repaint
    /// every cell (first frame, resize, or after a redraw request).
    pub fn encode(
        self: *Encoder,
        prev: *const FrameBuffer,
        next: *const FrameBuffer,
        full: bool,
    ) !EncodeResult {
        self.parts.clearRetainingCapacity();
        var changed: usize = 0;
        var dirty_rows: usize = 0;

        const w = next.width;
        const h = next.height;
        const same_size = prev.width == w and prev.height == h;
        const repaint = full or !same_size;
        if (repaint) try self.invalidateState();

        // Whether a cell actually changed. When the sizes disagree the previous
        // frame cannot be indexed with this frame's stride, so every cell counts
        // as different.
        const differs = struct {
            fn f(p: *const FrameBuffer, n: *const FrameBuffer, same: bool, i: usize) bool {
                if (!same) return true;
                return n.chars[i] != p.chars[i] or
                    n.fg[i] != p.fg[i] or
                    n.bg[i] != p.bg[i] or
                    !n.attrs[i].eql(p.attrs[i]);
            }
        }.f;

        for (0..h) |y| {
            const row_start = y * w;
            var x: usize = 0;
            var row_dirty = false;

            while (x < w) {
                // A full repaint forces every cell out even when unchanged;
                // `changed_cells` still reports real churn, not repaint volume,
                // which is why the two tests are separate.
                if (!(repaint or differs(prev, next, same_size, row_start + x))) {
                    x += 1;
                    continue;
                }

                // Walk left onto the lead cell if we landed on a wide char's tail.
                var start = x;
                while (start > 0 and next.chars[row_start + start] == unicode.continuation) {
                    start -= 1;
                }

                // Extend the run while cells are dirty, tolerating short gaps.
                var end = start;
                var clean: usize = 0;
                var probe = start;
                while (probe < w) : (probe += 1) {
                    if (repaint or differs(prev, next, same_size, row_start + probe)) {
                        end = probe;
                        clean = 0;
                    } else {
                        clean += 1;
                        if (clean > gap_merge) break;
                    }
                }

                try self.moveCursor(start, y);
                var cx = start;
                while (cx <= end) : (cx += 1) {
                    const j = row_start + cx;
                    const value = next.chars[j];
                    if (value == unicode.continuation) continue; // with its lead cell
                    try self.applyStyle(next.fg[j], next.bg[j], next.attrs[j]);
                    if (value == 0) {
                        try self.parts.append(self.allocator, ' ');
                    } else {
                        var buf: [4]u8 = undefined;
                        try self.parts.appendSlice(self.allocator, unicode.cellText(value, &buf));
                    }
                    self.state.x += @intCast(@max(1, unicode.cellWidth(value)));
                    if (differs(prev, next, same_size, j)) changed += 1;
                }
                // Writing the final column may have triggered autowrap; stop
                // trusting x.
                if (self.state.x >= @as(i64, @intCast(w))) self.state.known = false;
                row_dirty = true;
                x = end + 1;
            }
            if (row_dirty) dirty_rows += 1;
        }

        return .{
            .output = self.parts.items,
            .changed_cells = changed,
            .dirty_rows = dirty_rows,
        };
    }
};

/// One-shot encode of a whole buffer, e.g. for a screenshot. The caller owns the
/// returned bytes.
pub fn encodeFull(
    allocator: std.mem.Allocator,
    buf: *const FrameBuffer,
    options: Options,
) ![]u8 {
    var empty = try FrameBuffer.init(allocator, buf.width, buf.height);
    defer empty.deinit();
    var encoder = Encoder.init(allocator, options);
    defer encoder.deinit();
    const result = try encoder.encode(&empty, buf, true);
    return allocator.dupe(u8, result.output);
}

File: ports/rust/src/diff.rs
//! Turns two framebuffers into the smallest practical stream of escape
//! sequences. The encoder keeps a model of the terminal's current pen so no
//! redundant SGR is emitted.

use crate::ansi;
use crate::buffer::{Attrs, FrameBuffer};
use crate::capabilities::ColorDepth;
use crate::color::Color;
use crate::unicode::{cell_text, cell_width, CONTINUATION};

/// Rewriting up to this many unchanged cells is cheaper than the escape
/// sequence needed to jump over them, so neighbouring dirty spans get merged.
const GAP_MERGE: usize = 5;

#[derive(Clone, Copy)]
struct TerminalState {
    x: i64,
    y: i64,
    /// False after anything that makes the cursor position unknowable.
    known: bool,
    fg: Color,
    bg: Color,
    attrs: Attrs,
}

impl Default for TerminalState {
    fn default() -> TerminalState {
        TerminalState {
            x: -1,
            y: -1,
            known: false,
            fg: Color::DEFAULT,
            bg: Color::DEFAULT,
            attrs: Attrs::NONE,
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct EncodeResult {
    /// The bytes to hand to stdout.
    pub output: String,
    pub changed_cells: usize,
    /// Rows that contained at least one change.
    pub dirty_rows: usize,
}

#[derive(Clone, Copy, Debug, Default)]
pub struct EncoderOptions {
    pub colors: Option<ColorDepth>,
    /// Drain all color, keeping attributes.
    pub monochrome: bool,
}

pub struct Encoder {
    state: TerminalState,
    parts: String,
    pub colors: ColorDepth,
    pub monochrome: bool,
}

impl Default for Encoder {
    fn default() -> Encoder {
        Encoder::new(EncoderOptions::default())
    }
}

impl Encoder {
    pub fn new(options: EncoderOptions) -> Encoder {
        Encoder {
            state: TerminalState::default(),
            parts: String::new(),
            colors: options.colors.unwrap_or(ColorDepth::TrueColor),
            monochrome: options.monochrome,
        }
    }

    /// Forget what we believe about the terminal; the next write re-states
    /// everything.
    pub fn invalidate_state(&mut self) {
        self.state = TerminalState::default();
        self.parts.push_str("\x1b[0m");
    }

    fn fg_seq(&self, c: Color) -> String {
        if self.colors == ColorDepth::None {
            return String::new();
        }
        if c.is_default() {
            return ansi::FG_DEFAULT.to_string();
        }
        let col = if self.monochrome { c.grayscale() } else { c };
        match self.colors {
            ColorDepth::TrueColor => ansi::fg_true(col.red(), col.green(), col.blue()),
            ColorDepth::Ansi256 => ansi::fg_256(col.to_256()),
            _ => ansi::fg_16(col.to_16()),
        }
    }

    fn bg_seq(&self, c: Color) -> String {
        if self.colors == ColorDepth::None {
            return String::new();
        }
        if c.is_default() {
            return ansi::BG_DEFAULT.to_string();
        }
        let col = if self.monochrome { c.grayscale() } else { c };
        match self.colors {
            ColorDepth::TrueColor => ansi::bg_true(col.red(), col.green(), col.blue()),
            ColorDepth::Ansi256 => ansi::bg_256(col.to_256()),
            _ => ansi::bg_16(col.to_16()),
        }
    }

    fn apply_style(&mut self, fg: Color, bg: Color, attrs: Attrs) {
        if self.state.fg == fg && self.state.bg == bg && self.state.attrs == attrs {
            return;
        }

        // Attributes can only be added cheaply; removing one means a full reset.
        let removed = self.state.attrs & !attrs;
        if !removed.is_empty() {
            self.parts.push_str("\x1b[0m");
            self.state.attrs = Attrs::NONE;
            self.state.fg = Color::DEFAULT;
            self.state.bg = Color::DEFAULT;
        }

        let added = attrs & !self.state.attrs;
        if !added.is_empty() {
            let mut codes: Vec<u8> = Vec::new();
            if added.contains(Attrs::BOLD) {
                codes.push(1);
            }
            if added.contains(Attrs::DIM) {
                codes.push(2);
            }
            if added.contains(Attrs::ITALIC) {
                codes.push(3);
            }
            if added.contains(Attrs::UNDERLINE) {
                codes.push(4);
            }
            if added.contains(Attrs::BLINK) {
                codes.push(5);
            }
            if added.contains(Attrs::REVERSE) {
                codes.push(7);
            }
            if added.contains(Attrs::STRIKE) {
                codes.push(9);
            }
            if !codes.is_empty() {
                self.parts.push_str("\x1b[");
                for (i, code) in codes.iter().enumerate() {
                    if i > 0 {
                        self.parts.push(';');
                    }
                    self.parts.push_str(&code.to_string());
                }
                self.parts.push('m');
            }
            self.state.attrs = attrs;
        }

        if self.state.fg != fg {
            let seq = self.fg_seq(fg);
            self.parts.push_str(&seq);
            self.state.fg = fg;
        }
        if self.state.bg != bg {
            let seq = self.bg_seq(bg);
            self.parts.push_str(&seq);
            self.state.bg = bg;
        }
    }

    fn move_cursor(&mut self, x: usize, y: usize) {
        let (xi, yi) = (x as i64, y as i64);
        if self.state.known && self.state.y == yi {
            if self.state.x == xi {
                return;
            }
            if xi > self.state.x && xi - self.state.x <= 3 {
                // Short hop: cheaper than a full CUP, and never repaints cells.
                self.parts.push_str(&ansi::move_right((xi - self.state.x) as usize));
            } else if x == 0 {
                self.parts.push('\r');
            } else {
                self.parts.push_str(&ansi::move_to_column(x));
            }
        } else {
            self.parts.push_str(&ansi::move_to(x, y));
        }
        self.state.x = xi;
        self.state.y = yi;
        self.state.known = true;
    }

    /// Encode the difference between `prev` and `next`. Pass `full` to repaint
    /// every cell (first frame, resize, or after a redraw request).
    pub fn encode(&mut self, prev: &FrameBuffer, next: &FrameBuffer, full: bool) -> EncodeResult {
        self.parts.clear();
        let mut changed = 0usize;
        let mut dirty_rows = 0usize;

        let w = next.width;
        let h = next.height;
        let same_size = prev.width == w && prev.height == h;
        let repaint = full || !same_size;
        if repaint {
            self.invalidate_state();
        }

        // Whether this cell actually differs from the previous frame. When the
        // sizes disagree the previous frame cannot be indexed with this frame's
        // stride, so every cell counts as different.
        let differs = |i: usize| -> bool {
            if !same_size {
                return true;
            }
            next.chars[i] != prev.chars[i]
                || next.fg[i] != prev.fg[i]
                || next.bg[i] != prev.bg[i]
                || next.attrs[i] != prev.attrs[i]
        };
        // Whether this cell needs emitting, which a full repaint forces even
        // for cells that are unchanged. The two are deliberately separate:
        // `changed_cells` reports real churn, not repaint volume.
        let dirty_at = |i: usize| -> bool { repaint || differs(i) };

        for y in 0..h {
            let row_start = y * w;
            let mut x = 0usize;
            let mut row_dirty = false;

            while x < w {
                if !dirty_at(row_start + x) {
                    x += 1;
                    continue;
                }

                // Walk left onto the lead cell if we landed on a wide char's tail.
                let mut start = x;
                while start > 0 && next.chars[row_start + start] == CONTINUATION {
                    start -= 1;
                }

                // Extend the run while cells are dirty, tolerating short gaps.
                let mut end = start;
                let mut clean = 0usize;
                let mut probe = start;
                while probe < w {
                    if dirty_at(row_start + probe) {
                        end = probe;
                        clean = 0;
                    } else {
                        clean += 1;
                        if clean > GAP_MERGE {
                            break;
                        }
                    }
                    probe += 1;
                }

                self.move_cursor(start, y);
                for cx in start..=end {
                    let j = row_start + cx;
                    let value = next.chars[j];
                    if value == CONTINUATION {
                        continue; // emitted with its lead cell
                    }
                    self.apply_style(next.fg[j], next.bg[j], next.attrs[j]);
                    if value == 0 {
                        self.parts.push(' ');
                    } else {
                        self.parts.push_str(&cell_text(value));
                    }
                    self.state.x += cell_width(value).max(1) as i64;
                    if differs(j) {
                        changed += 1;
                    }
                }
                // Writing the final column may have triggered autowrap; stop
                // trusting x.
                if self.state.x >= w as i64 {
                    self.state.known = false;
                }
                row_dirty = true;
                x = end + 1;
            }
            if row_dirty {
                dirty_rows += 1;
            }
        }

        EncodeResult {
            output: std::mem::take(&mut self.parts),
            changed_cells: changed,
            dirty_rows,
        }
    }
}

/// One-shot encode of a whole buffer, e.g. for rendering a screenshot.
pub fn encode_full(buffer: &FrameBuffer, options: EncoderOptions) -> String {
    let empty = FrameBuffer::new(buffer.width, buffer.height);
    Encoder::new(options).encode(&empty, buffer, true).output
}
